scipy/optimize/_zeros_py.py

```diff
@@ -1393,3 +1393,328 @@ def toms748(f, a, b, args=(), k=1,
     x, function_calls, iterations, flag = result
     return _results_select(full_output, (x, function_calls, iterations, flag),
                            "toms748")
+
+
+def muller(f, x0, x1, x2, args=(), tol=1.48e-8, maxiter=50,
+           rtol=0.0, full_output=False, disp=True):
+    """
+    Find a root of a real or complex function using the Muller's method.
+
+    Find a root of the scalar-valued function `func` given a nearby scalar
+    starting points `x0`, `x1`, and `x2`.
+    The Muller method requires three starting points, but no derivatives,
+    and it works well for complex-valued roots. It will converge towards
+    complex roots even with real starting points. As with other root-finding
+    methods, initial guesses should be distinct.
```

Comment:
Is there a way to restrict it to real roots, e.g. if the user were interested in the only real root of a third order polynomial? (probably not?)

I'm wondering what happens when the callable doesn't support complex input/output. Basically, can we preserve the dtype of the user's input throught the calculation?

Reply:
>Is there a way to restrict it to real roots

I don't think so. That particular example would probably better be served with `secant`.

>I'm wondering what happens when the callable doesn't support complex input/output.

On the top of my head, I suspect that numpy (or similar) would throw a warning when casting complex to float.  Beyond that, we could add dtype checks for the first 3 function calls.

>Basically, can we preserve the dtype of the user's input throught the calculation?

I'd have to review the paper to see what type of functions it works for (it should work for any analytic function).  I can immediately make it fail to converge with `np.abs` and converge to nonsense values with `muller(lambda x: np.real((x-1)*(x+1+1j)), 2,5,3j, maxiter=100, full_output=True)`, so it's possible that artificially limiting it to reals would not work.

Reply:
OK, so what do you think about requiring that `x0`, `x1`, and `x2` be complex?